<a href="https://colab.research.google.com/github/Nov05/DS-Unit-2-Tree-Ensembles/blob/master/module2-random-forests/kaggle_20190715.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
# created by nov05 on 2019-07-15
# Kaggle
# DS-Unit-2-Tree-Ensembles/
# module1-decision-trees/
# module2-random-forests/
competition_name = "ds5-predictive-modeling-challenge"

https://www.kaggle.com/c/ds5-predictive-modeling-challenge/leaderboard  

In [0]:
from google.colab import drive
drive.mount('/content/drive')

In [4]:
# do this only once unless you create a new API key:
# create a folder with name "config" in "My Drive" 
# and store your API config file "kaggle.json" there

# do this every time after you reset all your runtimes,
# or kaggle won't find your API config file "kaggle.json" 
%env KAGGLE_CONFIG_DIR=/content/drive/My Drive/config

env: KAGGLE_CONFIG_DIR=/content/drive/My Drive/config


In [0]:
# run this code block to download competition files
!kaggle competitions download -c ds5-predictive-modeling-challenge
# unzip all .zip files
!unzip \*.zip

# Imports

In [0]:
!pip install --upgrade pandas-profiling
!pip install category_encoders

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pandas_profiling
import datetime as dt

import sklearn
from sklearn import datasets
from sklearn import linear_model
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier, AdaBoostRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_absolute_error, \
     mean_squared_error, r2_score
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split, \
     GridSearchCV
from sklearn.decomposition import PCA, TruncatedSVD

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, \
     OneHotEncoder, OrdinalEncoder, LabelEncoder

import category_encoders as ce
from category_encoders.binary import BinaryEncoder

%matplotlib inline

In [8]:
print("pandas_profiling", pandas_profiling.__version__)
print("category_encoders", ce.__version__)
print("pandas", pd.__version__)
print("sklearn", sklearn.__version__)

pandas_profiling 2.1.2
category_encoders 2.0.0
pandas 0.24.2
sklearn 0.21.2


# Download data

In [0]:
LOCAL = '../data/tanzania/'
# WEB = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Linear-Models/master/data/tanzania/'
WEB = '' # if you have downloaded the files to Colab folder '/content'

train_features = pd.read_csv(WEB + 'train_features.csv')
train_labels = pd.read_csv(WEB + 'train_labels.csv')
test_features = pd.read_csv(WEB + 'test_features.csv')
sample_submission = pd.read_csv(WEB + 'sample_submission.csv')

assert train_features.shape == (59400, 40)
assert train_labels.shape == (59400, 2)
assert test_features.shape == (14358, 40)
assert sample_submission.shape == (14358, 2)

In [79]:
X_train = train_features
y_train = train_labels['status_group']

X_train, X_val, y_train, y_val = train_test_split(
    X_train, y_train, train_size=0.80, test_size=0.20,
    stratify=y_train, random_state=42)

X_train.shape, X_val.shape, y_train.shape, y_val.shape

((47520, 40), (11880, 40), (47520,), (11880,))

# Feature examination/engineering

In [57]:
X_train['date_recorded'].describe()

count          47520
unique           349
top       2011-03-17
freq             474
Name: date_recorded, dtype: object

In [68]:
X_train.select_dtypes(include=np.number).columns

Index(['id', 'amount_tsh', 'gps_height', 'longitude', 'latitude',
       'num_private', 'region_code', 'district_code', 'population',
       'construction_year'],
      dtype='object')

In [0]:
# X_train['month_recorded_'] = pd.to_datetime(X_train['date_recorded'], infer_datetime_format=True)

In [103]:
X_train['recorded_by'].value_counts()

GeoData Consultants Ltd    47520
Name: recorded_by, dtype: int64

In [114]:
X_train[['water_quality', 'quality_group', 'quantity', 'quantity_group', 'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group']].describe()

water_quality quality_group  ...     waterpoint_type waterpoint_type_group
count          47520         47520  ...               47520                 47520
unique             8             6  ...                   7                     6
top             soft          good  ...  communal standpipe    communal standpipe
freq           40598         40598  ...               22778                 27642

[4 rows x 9 columns]

In [122]:
X_train[['extraction_type', 'extraction_type_group', 'extraction_type_class']].describe()

extraction_type extraction_type_group extraction_type_class
count            47520                 47520                 47520
unique              18                    13                     7
top            gravity               gravity               gravity
freq             21448                 21448                 21448

In [139]:
print(X_train['region'].value_counts().shape[0])
print(X_train['region_code'].value_counts().shape[0])
X_train[['region', 'region_code']].head()

21
27


region  region_code
43360   Mbeya           12
7263   Iringa           11
2486    Pwani            6
313    Kagera           18
52726  Dodoma            1

In [155]:
len(X_train['num_private'].value_counts())

59

In [102]:
print(X_train.shape[1])
X_train.columns

40


Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga',
       'ward', 'population', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'construction_year',
       'extraction_type', 'extraction_type_group', 'extraction_type_class',
       'management', 'management_group', 'payment', 'payment_type',
       'water_quality', 'quality_group', 'quantity', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group'],
      dtype='object')

In [0]:
# remove some features
features = X_train.columns.to_list()
features.remove('quality_group')
features.remove('quantity_group')
features.remove('source_type')
features.remove('source_class')
features.remove('waterpoint_type_group')
features.remove('extraction_type_group')
features.remove('extraction_type_class')
features.remove('id')
features.remove('date_recorded')
features.remove('region')

In [141]:
print(len(features))

30


# Preprocessing

In [148]:
numeric_features = X_train[features].select_dtypes(include=np.number).columns.to_list()
# categorical_features = X_train.select_dtypes(exclude=np.number).columns.to_list()
categorical_features, boolean_features = [], []
for col in X_train[features].select_dtypes(exclude=np.number):
  if isinstance(X_train[col][0], bool):
    boolean_features.append(col)
  else:
    categorical_features.append(col)
len(numeric_features), len(categorical_features), len(boolean_features)

(9, 19, 2)

In [149]:
# select high cardinality features
# for ordinal encoding (rather than one-hot encoding)
highcardi_features, lowcardi_features = [], []
for col in X_train[categorical_features]:
  if len(X_train[col].value_counts()) >= 10:
    highcardi_features.append(col)
  elif len(X_train[col].value_counts()) >= 2:
    lowcardi_features.append(col)
len(numeric_features), len(boolean_features), len(highcardi_features), len(lowcardi_features)

(9, 2, 11, 7)

In [0]:
f = 'region_code'
if f in numeric_features:
  numeric_features.remove(f)
  highcardi_features.append(f)

# Modeling

In [151]:
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())])
onehot_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(sparse=True, handle_unknown='ignore'))]) # sparse=False would exaust the memory
ordinal_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('ordinal', ce.OrdinalEncoder())])
boolean_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent'))])
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('boo', boolean_transformer, boolean_features),
        ('one', onehot_transformer, lowcardi_features), # categorical_features
        ('ord', ordinal_transformer, highcardi_features),
    ])

boost = AdaBoostClassifier(base_estimator=DecisionTreeClassifier())
parameters = {'n_estimators': (1, 50),
              'base_estimator__max_depth': (1, 26),
              'algorithm': ('SAMME', 'SAMME.R')}

clf = Pipeline(steps=[('preprocessor', preprocessor),
#                       ('pca', PCA(n_components=5, whiten=True)),
#                       ('svd', TruncatedSVD(n_components=200, random_state=5)), # 200059 cols
#                       ('lr', LogisticRegression(solver='lbfgs', multi_class='auto', n_jobs=-1)),
#                       ('dt', DecisionTreeClassifier(max_depth=26)),
#                       ('gscv', GridSearchCV(boost, parameters)) # might take long time
                      ('rf', RandomForestClassifier(n_estimators=200, random_state=5))
                     ])
clf.fit(X_train, y_train)

# print('TruncatedSVD explained variance ratio:', clf['svd'].explained_variance_ratio_.sum())
# print("Decision Tree feature importances:", clf['dt'].feature_importances_)
print("Validation score: %.3f" % clf.score(X_val, y_val))

Validation score: 0.811


rf, ordinal_transformer  
Validation score: 0.811 (all features)     
Validation score: 0.811 (removed some features)  
Validation score: 0.807 (removed more features)  
removed "region", ordinal encoded "region_code"  
Validation score: 0.811  
<br>  
('rf', RandomForestClassifier(n_estimators=100, random_state=5))  
Validation score: 0.810  
('rf', RandomForestClassifier())  
Validation score: 0.792  
('rf', RandomForestClassifier(n_estimators=100, max_depth=26, random_state=5))  
Validation score: 0.693  
<br>
('dt', DecisionTreeClassifier())   # depth=142  
Validation score: 0.780  
('dt', DecisionTreeClassifier(max_depth=75))  
Validation score: 0.770  
('dt', DecisionTreeClassifier(max_depth=15))  
Validation score: 0.715   
max_depth=29,30 / 0.782  
max_depth=28 / 0.784  
max_depth=27 / 0.786  
max_depth=26 / 0.782  
<br>
boost = AdaBoostClassifier(base_estimator=DecisionTreeClassifier())  
('gscv', GridSearchCV(boost, parameters))  
alidation score: 0.694  
<br>
('svd', TruncatedSVD(n_components=2000, random_state=5)), # 200059  
TruncatedSVD explained variance ratio: 0.8524638115439054  
Validation score: 0.773


In [0]:
# print(len(clf['dt'].feature_importances_))
# print(clf['dt'].max_features_)
# print(clf['dt'].get_depth())
# print(clf['dt'].feature_importances_[:15])

# Submission

In [156]:
test_pred = clf.predict(test_features)
submission = sample_submission.copy()
submission['status_group'] = test_pred
submission_filename = 'submission.csv'
submission.to_csv(submission_filename, index=False)
print(submission.shape)

(14358, 2)


In [157]:
# !kaggle competitions submit ds5-predictive-modeling-challenge -f submission.csv -m "random forest removed some features n_estimators=200 81.1%"

100% 263k/263k [00:04<00:00, 59.6kB/s]
Successfully submitted to DS5 Predictive Modeling Challenge

Adopting a Hypothesis-Driven Modeling Workflow  
https://outline.com/5S5tsB  

# Random Forest

https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html  

# AdaBoostClassifier

https://programtalk.com/python-examples/sklearn.tree.DecisionTreeRegressor/  
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.AdaBoostRegressor.html  
https://www.datacamp.com/community/tutorials/adaboost-classifier-python  

In [0]:
# iris = datasets.load_iris()
# boston = datasets.load_boston()

# def test_gridsearch():
#     # Check that base trees can be grid-searched.
#     # AdaBoost classification
#     boost = AdaBoostClassifier(base_estimator=DecisionTreeClassifier())
#     parameters = {'n_estimators': (1, 2),
#                   'base_estimator__max_depth': (1, 2),
#                   'algorithm': ('SAMME', 'SAMME.R')}
#     clf = GridSearchCV(boost, parameters)
#     clf.fit(iris.data, iris.target)
 
#     # AdaBoost regression
#     boost = AdaBoostRegressor(base_estimator=DecisionTreeRegressor(),
#                               random_state=0)
#     parameters = {'n_estimators': (1, 2),
#                   'base_estimator__max_depth': (1, 2)}
#     clf = GridSearchCV(boost, parameters)
#     clf.fit(boston.data, boston.target)

# test_gridsearch()

# TPOT

http://epistasislab.github.io/tpot/  